src/util.rs

```diff
@@ -62,6 +62,29 @@ where
 {
 }
 
+pub trait PushSimple<P> {
+    fn push_simple(&mut self, other: P);
+}
+
+impl PushSimple<PathBuf> for PathBuf {
+    fn push_simple(&mut self, other: PathBuf) {
+        for comp in other.components() {
+            match comp.as_os_str().to_str() {
```

Comment:
Why do you need to enforce valid UTF-8 here? You should be able to check the component like
```
if comp.as_os_str() == OsStr::new("..") { ... }
```